### ISSUE TITLE
/Messaging should take a single format in and out, look up provider by number

### ISSUE THREAD

**Issue by thefinn93** [AccelerateNetworks/NumberSearch#322]
I'm not sure of the current state of the stuff in the `Messaging` folder, but what I'd like to see:
* tracking of downstream clients (like orgsms), including API key and webhook URL for incoming messages
* an endpoint for those clients to send messages. This needs to be provider agnostic, including the request body. Look up the correct provider by source phone number and reformat the request to their specifications.
* broadcast incoming texts to each client's webhook URL. If you get a failure, backoff/retry for a while. Format here should also be consistent across providers.

basically want to move all of the logic of tracking providers and implementing per-provider APIs into a single place, and it seems like `Messaging` is the place to do it. Happy to chat about it.

![image](https://user-images.githubusercontent.com/692970/225433812-67a557ea-df76-4b4e-bab6-e05d8d1aa1e7.png)

**Comment by uncheckederror** [comment]
#309 is tracking the integration of FirstPointCom's SMS/MMS endpoints into the NumberSearch.Messaging project and the deployment of the Messaging app. It will have to be resolved first before we're able to complete this.

**Comment by uncheckederror** [comment]
For a single message from the PoV of the user, where there is a sentence of text and a pair of images, it appears that FirstPointCom would expect us to send them an SMS request containing the text, and then one or more MMS requests containing byte array encoded files. This is interesting because it a multi-step process but also because BulkVS required an array of strings called MediaURLs to send an MMS. Where each MediaURL string was an HTTP route to some file. These are two very different approaches, so I am not sure what would be the best approach to extend the existing generic /SendMessage endpoint to support MMS messages.

Especially given the current send message format is rather simple: 
{
  "to": "string",
  "msisdn": "string",
  "message": "string"
}

To extend this directly for supporting MMS through FirstPointCom something like this could work:

{
  "to": "string",
  "msisdn": "string",
  "message": "string"
  "files" : [ "Base64EncodedFileFromBytes1", "Base64EncodedFileFromBytes2" ]
}

And then in the sms.callpipe.com API I can break this out an SMS request for the "message" field, and then 2 MMS requests, one for each of the images as they only support sending one file at a time. Where this API takes the base64 formatted files and converts them to a byte[] which is the format that the vendor's MMS API expects. Byte arrays are not directly supported in the JSON format, from what I've read everyone handles this by base64 encoding their byte array and then listing it as a string field inside the JSON body of the message, hence why I am doing that here.

---

Here's another thing I was considering; Is "msisdn" a good field name? We could just rename it to "from", although I am not sure which name is more useful as a description.

**Comment by danry25** [comment]
@uncheckederror Re-reading Harel's examples, I'm pretty sure they expect a .txt file containing the MMS text content when sending a MMS. Not sure how group MMS sending would work, I believe I remember seeing on the incoming endpoint we receive a list of numbers in the MMS group?

{"sessionid": "redo 07:45:15:2277e10b9f5ad6c407c9877e509d0da5fa4", "origtime": "2022-04-24 06:39:00", "timezone": "EST", "to": "12123334444", "msisdn": "12121234567", "message": “{ "authkey":"l23r324k12l4l1;", "encoding":"native",
"files":"part-001.jpg,part-002.txt,", "recip":"+12123334444,",
"url":"https://mmsc01.1pcom.net/MMS_Pickup?msgid=7e10b9f5ad6cc07c987Fe509d0da5frw
}", "serversecret": "secrettoken", "api_version": 0.5}

**Comment by uncheckederror** [comment]
@danry25 in his example for sending an MMS he just passes the to, from, and a file as bytes
`-F username=yourapikey -F password=yourpassword -F recip=12125555555 -F ani=12123334444 -F [ufiles=@/tmp/image001.png](mailto:ufiles=@/tmp/image001.png)`

The JSON body you posted is what gets forwarded from their webhook to our sms.callpipe.com API.

**Comment by thefinn93** [comment]
it looks like the field name is "ufiles" (plural) implying multiple files could be sent. What if we send a .txt file with the message along with the actual attachment?

**Comment by uncheckederror** [comment]
I have updated the JSON that is forwarded to registered clients based on @thefinn93 feedback on what his client apps are actually consuming.

Here is the new structure as described in the live docs:

```
{
  "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  "from": "string",
  "to": [
    "string"
  ],
  "content": "string",
  "mediaURLs": [
    "string"
  ],
  "messageType": "SMS",
  "dateReceivedUTC": "2023-05-09T06:43:08.111Z",
  "clientSecret": "string"
}
```

Based on the incoming MMS work I have been doing I would like to propose that we update the /message/send endpoint to accept a JSON body in this format:

```
{
    "to": [ "string", "string" ]
    "msisdn": "string",
    "message": "string"
    "mediaURLs" : [ "string", "string" ]
}
```

Where the `to` field is a string array now rather than just a string to enhance support for sending messages to groups. Additionally a `mediaURLs` field has been added that is another string array. The URLs for the files stored in our DO Space would be listed here so that the sms.callpipe.com API app could download and forward them to our upstream vendors.

**Comment by uncheckederror** [comment]
@thefinn93 okay based on you feedback here's the updated format for forwarded messages:

```
{
  "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  "from": "string",
  "to": "string",
  "additionalRecipients": [
    "string"
  ],
  "content": "string",
  "mediaURLs": [
    "string"
  ],
  "messageType": "SMS",
  "dateReceivedUTC": "2023-05-10T03:11:38.545Z",
  "clientSecret": "string"
}
```

Where To is a single dialed number that is the registered client that the message is being forwarded to.

For sending outbound messages, both SMS and MMS, the format has been updated to this:

```
{
  "to": "string",
  "msisdn": "string",
  "message": "string",
  "mediaURLs": [
    "string"
  ]
}
```

To better support MMS messages.

**Comment by thefinn93** [comment]
how do I send to a group?

**Comment by uncheckederror** [comment]
The To field can be a single dialed number or multiple comma separated dialed numbers. For example:

```
{
  "to": "number1,number2,number3",
  "msisdn": "fromNumber",
  "message": "ayyyoooo!",
  "mediaURLs": [
    "fileToDownload.jpeg",
    "fileToRead.txt"
  ]
}
```

**Comment by thefinn93** [comment]
that works. Thanks

**Comment by danry25** [comment]
@thefinn93 Are you happy with where we are at on the SMS and MMS APIs provided by this application for OrgSMS and Web Texting's consumption? If so, please close this issue.
